### DISCUSSION TITLE
Nested query - How to fetch categories of each product

### DISCUSSION THREAD

**Discussion by agsmartwolf** [swellstores/community#230] [Help] [answered]
Hi, I've added permission and there is only one 'id 'field in response right now. The production environment is effected.

Request is:
<img width="832" alt="Screenshot 2023-04-24 at 19 14 40" src="https://user-images.githubusercontent.com/109105178/234040259-9c1f8bfa-beb2-4cb2-baa5-668b555c1866.png">

response:
<img width="676" alt="Screenshot 2023-04-24 at 19 43 13" src="https://user-images.githubusercontent.com/109105178/234047857-5aa72e31-f42e-48e3-bf70-3b0d37605a31.png">

Permissions:
<img width="624" alt="image" src="https://user-images.githubusercontent.com/109105178/234048398-a109cf13-4eb0-476f-9ae0-7eebc5ad2812.png">

**Comment by agsmartwolf** [comment]
I have a category which includes all of the existed products. I use this approach I could sort products in the dashboard to display all of the products sorted in specific way.

**Comment by agsmartwolf** [comment]
Guys, do you have any solutions?

**Reply by ericingram** [reply]
@StevenSwell let's triage this

**Reply by steven-k1** [reply]
@agsmartwolf Will have our team look into it and advise once we have more details.

**Reply by agsmartwolf** [reply]
So problems are appearing if I add "products.categories", if I remove this permission, It works.

But I need to know what categories products has.

<img width="666" alt="Screenshot 2023-04-28 at 18 38 16" src="https://user-images.githubusercontent.com/109105178/235177791-dea4e803-5d7c-42d6-b89b-83f594fbf8ea.png">

**Reply by agsmartwolf** [reply]
Graphql client works fetching categories field. However, I can't use it because there is a bug that we can't use nested querying. So 15 products are always fetched for the category.

**Reply by agsmartwolf** [reply]
Could you please estimate periods when one of these problems would be resolved?
I have an old version of the site on production right now. I can't deploy any changes to production due to these bugs.

**Reply by logeshswell** [reply]
@agsmartwolf  The fix is to remove the `products.categories` permission which you did already and it should work fine now.

**Reply by agsmartwolf** [reply]
But I need this field in the response.

**Reply by agsmartwolf** [reply]
@logeshswell  I need to know what other categories can include this product.

**Reply by logeshswell** [reply]
@agsmartwolf I think you are using the wrong query. Inorder to help with this, may I know what you exactly want to achieve?

**Reply by logeshswell** [reply]
I can understand that you want to list the first 100 products present in the category.
Additionally, do you wish to
i) Fetch sub-categories of a category (or)
ii) For each product, fetch associated categories
iii) Anything else?

**Reply by agsmartwolf** [reply]
Yeah, of course. I want to display all of products in certain order. There were 2 ways to archive it - create a category which contains all of the products, or create a custom integer field and manually set order of products to display.

However I want also manager could order products by himself in the SWELL dashboard, so I've chosen the option with additional category which contains all existed products.

I need `categories` field in response, because I filter products by this field when user changes categories.
I already have all of products, stored on frontend side, so to send extra request to the SWELL API to get filtered products is not a good option for me. (API calls are also too slow - sometimes 4-7 seconds)

**Reply by logeshswell** [reply]
@agsmartwolf I get your point of sorting all the products by introducing a new category (All Products). But, I am still not clear with the last section. 
Do you wish to have,
i) Categories of each product filtered by `category_id` (or)
ii) Fetch all categories present in your store.

**Reply by agsmartwolf** [reply]
I want to have all categories for each product.
In the product model, there is a field `categories`, which includes all categories, which include the current product.

So I fetch a category and want something like this:
```
category:
    id: string
    slug: string
    ...
    products
        id: string
        ...
        categories: Array<{ name: string; slug: string }>
```

**Reply by logeshswell** [reply]
Understood @agsmartwolf. So, you need to use `"products.product.categories"` which would fetch all categories present in each product.

But, the above solution would work only in backend APIs as we need to open up some permissions for frontend API. 

I am not very sure if we can open up permissions for this scenario as it gets deeper into `expand` attribute instead of `fields`. But, I will look into it and confirm you in sometime.

**Reply by logeshswell** [reply, marked as answer]
I found an answer to your query. Please check and let me know if you face any other issues.

First you need to setup the permissions as follows:
```json
{
  "permissions": {
    "categories": {
      "fields": [
        "products.name",
        "products.slug",
        "products.price",
        "products.categories",
        "products.categories.parent"
      ],
      "expands": {
        "products": {
          "url": "/products",
          "params": {
            "categories": "id"
          },
          "data": {
            "expand": ["categories.parent"]
          }
        }
      }
    }
  }
}
```

Secondly, you can use swell-js to query as follows:
```ts
let categoryDetails = await swell.categories.get('624d956b8c83fa0133235e3f', {
  "expand": ["products"]
});
console.log(categoryDetails);
```

**Output:**
```json
{
    "sort": 0,
    "slug": "cosmetics",
    "name": "Cosmetics",
    "id": "624d956b8c83fa0133235e3f",
    "products": {
        "count": 1,
        "pageCount": 1,
        "page": 1,
        "results": [
            {
                "id": "61c75dda2280080132420500",
                "name": "Hand bag",
                "slug": "hand-bag",
                "price": 24.95,
                "categories": {
                    "count": 2,
                    "pageCount": 1,
                    "page": 1,
                    "results": [
                        {
                            "parentId": "624c937f9f1c73013cebda59",
                            "productId": "61c75ddb2280080132420b39",
                            "sort": 0,
                            "dateCreated": "2022-09-27T20:26:21.646Z",
                            "id": "63335c6d9796770012147a65",
                            "parent": {
                                "name": "Accessories",
                                "active": true,
                                "slug": "accessories",
                                "dateCreated": "2022-04-05T19:07:43.955Z",
                                "dateUpdated": "2023-01-25T23:41:39.772Z",
                                "sort": 1,
                                "id": "624c937f9f1c73013cebda59"
                            }
                        }
                        {
                            "parentId": "624d956b8c83fa0133235e3f",
                            "productId": "61c75dda2280080132420500",
                            "sort": 0,
                            "dateCreated": "2023-01-05T22:05:42.449Z",
                            "id": "63b749b69bd1a200125bb256",
                            "parent": {
                                "name": "Cosmetics",
                                "active": true,
                                "slug": "cosmetics",
                                "dateCreated": "2022-04-06T13:28:11.980Z",
                                "dateUpdated": "2022-04-06T15:10:49.731Z",
                                "sort": 0,
                                "id": "624d956b8c83fa0133235e3f"
                            }
                        }
                    ]
                }
            }
        ]
    }
}
```

**Reply by agsmartwolf** [reply]
@logeshswell 
Yeah, it works great, thank you! I've implemented all the logic I wanted with approach you suggested.

**Reply by agsmartwolf** [reply]
One more question: how to get localized response?
I've tried to add `locale` parameter, but `null` is returned.

**Reply by agsmartwolf** [reply]
@logeshswell

**Reply by agsmartwolf** [reply]
@logeshswell Could you please help me to get localized response?

**Reply by logeshswell** [reply]
@agsmartwolf Have you tried [localization](https://developers.swell.is/frontend-api/localization) following the below code?
```ts
await swell.locale.select("ru");
```
